src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs

```diff
@@ -52,8 +45,16 @@ private static void HandleStartArray(JsonSerializerOptions options, ref ReadStac
 
             state.Current.CollectionPropertyInitialized = true;
 
-            // We should not be processing custom converters here (converters are of ClassType.Value).
-            Debug.Assert(state.Current.JsonClassInfo.ClassType != ClassType.Value);
+            // The current JsonPropertyInfo will be null if the current type is not one of
+            // ClassType.Value | ClassType.Enumerable | ClassType.Dictionary.
+            // We should not see ClassType.Value here because we handle it on a different code
+            // path invoked in the main read loop.
+            // Only ClassType.Enumerable is valid here since we just saw a StartArray token.
+            if (state.Current.JsonPropertyInfo == null ||
+                (ClassType.Enumerable & state.Current.JsonPropertyInfo.ClassType) == 0)
+            {
+                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.JsonClassInfo.Type);
+            }
```

Comment:
Moved the check from https://github.com/dotnet/corefx/pull/42254/commits/a8888973b27b747cbb76bad106647920e4d63f16#diff-0cae46d8fb86e151bbde613cf263915fR189-R197 here for clearer context.

Reply:
Was it intentional to no longer pass `jsonPropertyInfo.RuntimePropertyType`?

Reply:
Yes. For simplicity of the logic, the class info's `Type` (which is the same as the `jsonPropertyInfo.DeclaredPropertyType`, if it exists) can be used. 

In a future PR (to keep this one as targeted as possible), it could be good to run through where we throw JsonException (couldn't convert value) and decide if it's better to display the runtime type or the declared type. I think declared is better because that is the input/given type.